snapshot/git/gitdb/bundlestore.go

```diff
@@ -0,0 +1,200 @@
+	revList := bundlestoreTempRef
+
+	// the name of the stream that this bundle requires
+	streamName := ""
+
+	switch s.kind {
+	case kindGitCommitSnapshot:
+		if db.stream.cfg != nil && db.stream.cfg.RefSpec != "" {
+			streamHead, err := db.dataRepo.Run("rev-parse", db.stream.cfg.RefSpec)
+			if err != nil {
+				return nil, err
+			}
+
+			mergeBase, err := db.dataRepo.RunSha("merge-base", streamHead, commitSha)
+
+			// if err != nil, it just means we don't have a merge-base
+			if err == nil {
+				revList = fmt.Sprintf("%s..%s", mergeBase, bundlestoreTempRef)
+				streamName = db.stream.cfg.Name
+			}
+
+		}
+	case kindFSSnapshot:
+		commitSha, err = db.dataRepo.RunSha("commit-tree", commitSha, "-m",
+			"commit to distribute GitDB FSSnapshot via bundlestore")
+	default:
+		return nil, fmt.Errorf("unknown Snapshot kind: %v", s.kind)
+	}
+
+	// update the ref
+	if _, err := db.dataRepo.Run("update-ref", bundlestoreTempRef, commitSha); err != nil {
+		return nil, err
+	}
+
+	d, err := db.tmp.TempDir("bundle-")
+	if err != nil {
+		return nil, err
+	}
+
+	// we can't use tmpDir.TempFile() because we need the file to not exist
+	bundleFilename := path.Join(d.Dir, fmt.Sprintf("commit-%s.bundle", s.SHA()))
+
+	// create the bundle
+	if _, err := db.dataRepo.Run("bundle", "create", bundleFilename, revList); err != nil {
+		return nil, err
+	}
+
+	f, err := os.Open(bundleFilename)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	if err := b.cfg.Store.Write(s.sha, f); err != nil {
+		return nil, err
+	}
+
+	return &bundlestoreSnapshot{sha: s.SHA(), kind: s.Kind(), bundleName: s.sha, streamName: streamName}, nil
+}
+
+type bundlestoreSnapshot struct {
+	sha        string
+	kind       snapshotKind
+	bundleName string
+	streamName string
+}
+
+func (s *bundlestoreSnapshot) ID() snap.ID {
+	return snap.ID(strings.Join([]string{bundlestoreIDText, string(s.kind), s.bundleName, s.streamName, s.sha}, "-"))
+}
+func (s *bundlestoreSnapshot) Kind() snapshotKind { return s.kind }
+func (s *bundlestoreSnapshot) SHA() string        { return s.sha }
+
+func (s *bundlestoreSnapshot) Download(db *DB) error {
+	if err := db.shaPresent(s.SHA()); err == nil {
+		return nil
+	}
+
+	// TODO(dbentley): keep stats about bundlestore downloading
+	// TODO(dbentley): keep stats about how long it takes to unbundle
+	filename, err := s.downloadBundle(db)
+	if err != nil {
+		return err
+	}
+
+	// unbundle optimistically
+	// this will succeed if we have all of the prerequisite objects
+	if _, err := db.dataRepo.Run("bundle", "unbundle", filename); err == nil {
+		return db.shaPresent(s.sha)
+	}
+
+	// we couldn't unbundle
+	// see if it's because we're missing prereqs
+	exitError := err.(*exec.ExitError)
+	if exitError == nil || !strings.Contains(string(exitError.Stderr), "error: Repository lacks these prerequisite commits:") {
+		return err
+	}
+
+	// we are missing prereqs, so let's try updating the stream that's the basis of the bundle
```

Comment:
or you're offline or....

The second unbundle would fail because the git server is out-of-date, or offline, or...

I've added some comments.

And yeah, I think this is a higher level of review and I'm not sure I can imagine everything right now.